sched/config/config.go

```diff
@@ -3,445 +3,199 @@ package config
-}
-type QueueConfig struct {
-	Type     QueueType `json:","`
-	Count    *int      `json:",omitempty"` //fake
-	Capacity *int      `json:",omitempty"` //memory
-}
-type SagaCoordConfig struct {
-	Type SagaCoordType `json:","`
-}
-type ExecConfig struct {
-	Type ExecType `json:","`
-}
-type RunnerConfig struct {
-	Type RunnerType `json:","`
-	Exec ExecConfig `json:","`
-}
-type WorkerConfig struct {
-	Type           WorkerType   `json:","`
-	Runner         RunnerConfig `json:","`
-	PingIntervalMs int          `json:","`
-	TimeoutMs      int          `json:","`
-}
-type PlannerConfig struct {
-	Type PlannerType `json:","`
-}
-type StatsConfig struct {
-	Path    string `json:","`
-	Enabled bool   `json:","`
-}
-type HealthConfig struct {
-	Path    string `json:","`
-	Ok      string `json:","`
-	Enabled bool   `json:","`
+// Generates the JSON config that results from the empty string; useful for showing a complete configuration.
+func (p *Parser) DefaultJSON() ([]byte, error) {
+	i, err := p.Parse(nil)
+	if err != nil {
+		return nil, err
+	}
+	return json.Marshal(i)
 }
 
-// Takes a subsection of the config file and constructs the appropriate type.
-type Plugins map[TypeId]PluginFn
-type PluginFn func(json.RawMessage) (interface{}, error)
-type PluginType string
-
-// All the types we construct based on the config.
-type TypeId string
-type ClusterType PluginType
-type SagaCoordType PluginType
-type WorkerType PluginType
-type QueueType PluginType
-type PlannerType PluginType
-type RunnerType string //PluginType
-type ExecType string   //PluginType
-
-const (
-	//
-	ClusterTypeId   TypeId = "cluster"
-	SagaCoordTypeId TypeId = "sagaCoord"
-	WorkerTypeId    TypeId = "worker"
-	QueueTypeId     TypeId = "queue"
-	PlannerTypeId   TypeId = "planner"
-	RunnerTypeId    TypeId = "runner"
-	ExecTypeId      TypeId = "exec"
-	StatsTypeId     TypeId = "stats"
-	HealthTypeId    TypeId = "health"
-
-	//
-	ClusterMemory  ClusterType = "memory"
-	ClusterStatic  ClusterType = "static"
-	ClusterDynamic ClusterType = "dynamic"
-
-	//
-	QueueFake    QueueType = "fake"
-	QueueMemory  QueueType = "memory"
-	QueueDurable QueueType = "durable"
-
-	//
-	SagaCoordMemory  SagaCoordType = "memory"
-	SagaCoordDurable SagaCoordType = "durable"
-
-	//
-	WorkerMemory WorkerType = "memory"
-	WorkerThrift WorkerType = "thrift"
-
-	//
-	PlannerChaos  PlannerType = "chaos"
-	PlannerSimple PlannerType = "simple"
-	PlannerSmart  PlannerType = "smart"
-
-	//
-	RunnerSimple RunnerType = "simple"
-
-	//
-	ExecOs  ExecType = "os"
-	ExecSim ExecType = "sim"
-)
-
-func IntToPtr(i int) *int { return &i }
-
-/*
-
-
-## Example json. Any subset can be specified to override specific defaults.
-##
+func (p *Parser) Parse(configText []byte) (*Config, error) {
+	// TODO(dbentley): allow this to refer to a file instead of being a giant string
+	if len(configText) == 0 {
+		configText = emptyJson
+	}
+	var cfg topLevelConfig
+	err := json.Unmarshal(configText, &cfg)
```

Comment:
Well, it's up to how you configure the parser. If you set "memory" to have a default...

Basically, this is split so there's a framework (config.go) and then a configuration of it (defaults.go)...